tests/test_odcs_requester.py

```diff
@@ -0,0 +1,83 @@
+"""test_odcs_requester.py - test odcs_requester"""
+from typing import Callable
+
+# pylint: disable=redefined-outer-name
+from unittest.mock import MagicMock, create_autospec
+
+import pytest
+from odcs.client.odcs import ODCS, ComposeSourceTag  # type: ignore
+
+from generate_compose.odcs_configurations_generator import (
+    ODCSComposeConfig,
+    ODCSComposesConfigs,
+)
+from generate_compose.odcs_requester import ODCSRequester, ODCSRequestReferences
+
+
+@pytest.fixture()
+def compose_url() -> str:
+    """Example compose-url, as close to the one expected"""
+    return "http://download.eng.bos.redhat.com/odcs/prod/odcs-222222"
+
+
+@pytest.fixture()
+def create_odcs_mock(compose_url: str) -> Callable[[int], MagicMock]:
+    """Create an ODCS mock with specific results for the compose method"""
+
+    def _mock_odcs_compose(num_of_composes: int) -> MagicMock:
+        """Monkey-patched ODCS compose"""
+        mock: MagicMock = create_autospec(ODCS)
+        mock.request_compose.side_effect = [
+            {"result_repofile": f"{compose_url}", "id": i}
+            for i in range(1, num_of_composes + 1)
+        ]
+        return mock
+
+    return _mock_odcs_compose
+
+
+@pytest.fixture()
+def compose_source_tag() -> ComposeSourceTag:
+    """Example ComposeSource of type 'tag'"""
+    return ComposeSourceTag(tag="tag")
+
+
+@pytest.mark.parametrize(
+    ("composes_configs, num_of_composes"),
+    [
+        pytest.param(
+            ODCSComposesConfigs([ODCSComposeConfig(spec=ComposeSourceTag(tag="tag"))]),
+            1,
+            id="single compose, tag source, no additional arguments",
+        ),
+        pytest.param(
+            ODCSComposesConfigs(
+                [
+                    ODCSComposeConfig(spec=ComposeSourceTag(tag="tag")),
+                    ODCSComposeConfig(spec=ComposeSourceTag(tag="tag")),
+                ]
+            ),
+            2,
+            id="multiple composes, , tag source, no additional arguments",
+        ),
+    ],
+)
+def test_odcs_requester(
+    create_odcs_mock: Callable,
+    compose_url: str,
+    compose_source_tag: ComposeSourceTag,
+    composes_configs: ODCSComposesConfigs,
+    num_of_composes: int,
+) -> None:
+    """test ODCSRequester.__call__"""
+    expected_req_ref = ODCSRequestReferences(
+        compose_urls=[compose_url] * num_of_composes
+    )
+    mock_odcs = create_odcs_mock(num_of_composes)
+    odcs_requester = ODCSRequester(odcs=mock_odcs)
+    req_ref = odcs_requester(compose_configs=composes_configs)
+
+    assert mock_odcs.request_compose.call_count == num_of_composes
+    assert mock_odcs.request_compose.call_args[0][0].source == compose_source_tag.source
```

Comment:
I think that this should be replaced with `call_arg_list` that captures all the calls to the mock object.
In addition, I think that for generating the expected calls, you should use `composes_configs`.
`compose_source_tag` can be removed from this test, I don't see why we need it and `compose_configs` at the same time.

Reply:
Thing is, when we call `request_compose` we give it a compose source as `spec`. So if we are using `call_args_list`, we will have to compare between the same objects (in memory), while we don't have access to the one created in the parametrization.  
we can instead either have separated tests for the different test-cases, or generate the objects in the test function according to the number of composes needed for each test, but I think it is not as good of a long-term solution (if we'd like to add test cases for different compose sources, or with different kwargs)

Reply:
replaced